recwav.1
.TH recwav 1 "May 2000" Linux "Linux Audio Manual"
.SH NAME
recwav \- WAV File Recording Tool
.SH SYNOPSIS
.B recwav [-v][-i][-l][-p <dsp-dev>] <wav-file>

.SH DESCRIPTION
\fBRecwav\fP is a command-line tool for recording audio files in the
Microsoft WAV format.

.SH OPTIONS
.TP
.B \-v
Print a version identifier and then exit.
.TP
.B \-t <length>
Make a recording <length> seconds long.  If no length is specified, the 
recording will continue until the INTERRUPT key (typically CNTL-C) is
pressed.
.TP
.B \-d <dsp-dev>
Use \fI<dsp-dev>\fP as the recording device.  If not specified, \fI/dev/dsp\fP
will be used by default.
.TP
.B \-r <samp-rate>
Record using a sampling rate of \fI<samp-rate>\fP samples per second.  If not 
specified, a sampling rate of 44,100 samples per second will be used
by default.
.TP
.B \-s <samp-size>
Record using a sample size of \fI<samp-size>\fP bits per sample.  If not
specified, a sample size of 16 bits per sample will be used by default.
.TP
.B \-c 1|2
Record using 1 or 2 channels.
.SH TODO
Currently, \fBrecwav\fP is capable of recording audio only in
Linear PCM format.  MPEG support would be a Good Thing.
.SH AUTHOR
Fred Gleason <fredg@wava.com>


History
                    History file for the WAVLIB Project

-------------------------------------------------------------------------------
11/20/1999
Start of history file

Current released version is 2.2.2

Added the aproc(1) command, a command-line audio processor for wav files.
Currently, it only supports the 'normalize' function (-N), but is intended
as a catch-all for adding additional processing features as needed through
addition of additional command flags.

Added a man page for the aproc(1) command.

Fixed a bug in OpenWav() that was causing the file pointer to not be
pointing at the start of the wave data at function return with certain 
combinations of chunks. 

Fixed a typo in the ameter(1) man page.
-------------------------------------------------------------------------------

12/1/1999 -- v 2.3.1
Added a conditional definition of PI (definition is missing from glibc).
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.0

Cleaned up code in ameter(1) so that levels are now displayed in true 
logarithmic dB.

Added a '-l' option to ameter(1) to allow the user to set the meter reference
level

Added color to ameter(1), with a green and red segment.  Added a '-t' option
to allow the user to define where the border should be.

Updated the ameter(1) man page.

Fixed a bug in infowav(1) that was causing it to corrupt wav files when invoked
with the '-p' while omitting the audio-device name that customarily follows it.

Fixed a bug in infowav(1) that was causing extraneous output to be printed.

Added a '-t' option to recwav(1) to allow the user to specify a record length.

Updated the recwav(1) man page.

Added a TailTrim() API call to remove audio below a given set threshold level
from the end of an existing wav file.
-------------------------------------------------------------------------------

5/18/2000 -- v 2.4.1

Restored the RecordWavFile() API call, which somehow got dropped out of
v2.4.0.
-------------------------------------------------------------------------------

5/23/2000 -- v 2.4.2

Fixed a fencepost bug in TailTrim() that was causing false positive 
threshold detects on certain wav files.
-------------------------------------------------------------------------------

recwav.c
/* recwave.c
 *
 *  A utility for recording files in the the Microsoft 'wav' format.
 *
 *   (C) Copyright 2002-2005 Fred Gleason <fredg@salemradiolabs.com>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as
 *   published by the Free Software Foundation; either version 2 of
 *   the License, or (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public
 *   License along with this program; if not, write to the Free Software
 *   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <signal.h>
#include "wavlib.h"
#define USAGE "recwave [-v][-d <dev>][-r <rate>][-s <size>][-c <chan>][-t <length>] <wav-file>"
#define DEFAULT_DSP "/dev/dsp"
#define DEFAULT_SAMPLERATE 44100
#define DEFAULT_SAMPLESIZE 16
#define DEFAULT_CHANNELS 2
#define REC_SAMPLES 882000

int dStop=0;
void stoprecord(int);

int main(int argc,char *argv[])
{
  int i,j;
  int hFilename,hAudio;
  struct wavHeader wavHeader;
  int dOptionFound;
  char sDspdev[256];
  unsigned short wChannels=DEFAULT_CHANNELS;
  unsigned dwSamplesPerSec=DEFAULT_SAMPLERATE;
  unsigned short wBitsPerSample=DEFAULT_SAMPLESIZE;
  unsigned dAccum;
  unsigned char aBuffer[AUDIO_BUFFER];
  unsigned dCount=0;
  unsigned dBytes,dSamples;
  unsigned dLength=0;

  /*
   * Set default DSP Device
   */
  strcpy(sDspdev,DEFAULT_DSP);

  /*
   * Parse Options
   */
  if(argc<2) {
    printf(USAGE);
    printf("\n");
    exit(1);
  }
  if(argc==2) {
    if(strcmp(argv[1],"-v")==0) {   /* Version */
      printf("recwav v");
      printf(VERNUM);
      printf("\n");
      exit(0);
    }
  }
  for(i=1;i<argc-1;i++) {
    dOptionFound=0;
    if(strcmp(argv[i],"-d")==0) {    /* Select DSP Device */
      strcpy(sDspdev,argv[++i]);
      dOptionFound=1;
    }
    if(strcmp(argv[i],"-r")==0) {    /* Select Sample Rate */
      if(sscanf(argv[++i],"%u",&dwSamplesPerSec)!=1) {
	printf("recwave: invalid sample rate\n");
	exit(1);
      }
      dOptionFound=1;
    }
    if(strcmp(argv[i],"-t")==0) {    /* Record Time */
      if(sscanf(argv[++i],"%u",&dLength)!=1) {
	printf("recwave: invalid record length\n");
	exit(1);
      }
      dOptionFound=1;
    }
    if(strcmp(argv[i],"-s")==0) {    /* Select Sample Size */
      if(sscanf(argv[++i],"%u",&dAccum)!=1) {
	printf("recwave: invalid sample size\n");
	exit(1);
      }
      if((dAccum!=8)&&(dAccum!=16)) {
	printf("recwave: invalid sample size\n");
	exit(1);
      }
      wBitsPerSample=(unsigned short)dAccum;
      dOptionFound=1;
    }
    if(strcmp(argv[i],"-c")==0) {    /* Select Number of Channels */
      if(sscanf(argv[++i],"%u",&dAccum)!=1) {
	printf("recwave: invalid channel argument\n");
	exit(1);
      }
      if((dAccum!=1)&&(dAccum!=2)) {
	printf("recwave: invalid channel argument\n");
	exit(1);
      }
      wChannels=(unsigned short)dAccum;
      dOptionFound=1;
    }
    if(dOptionFound==0) {
      printf("recwave: invalid option\n");
      exit(1);
    }
  }

  /*
   * If a length is specified, simply pass the appropriate arguments
   * to RecWavFile()
   */
  if(dLength>0) {
    printf("Recording for %u seconds\n",dLength);
    if(RecWavFile(argv[argc-1],sDspdev,dLength,wChannels,dwSamplesPerSec,
		  wBitsPerSample,0)<0) {
      perror("recwav");
      exit(1);
    }
    exit(0);
  }

  /*
   * Open DSP Device
   */
  hAudio=open(sDspdev,O_RDONLY);
  if(hAudio<0) {
    printf("recwave: can't open dsp device\n");
    exit(1);
  }

  /*
   * Open the wav file
   */
  wavHeader.wFormatTag=WAVE_FORMAT_PCM;
  wavHeader.wChannels=wChannels;
  wavHeader.dwSamplesPerSec=dwSamplesPerSec;
  wavHeader.dwAvgBytesPerSec=dwSamplesPerSec*(wBitsPerSample/8)*wChannels;
  wavHeader.wBlockAlign=(wBitsPerSample/8)*wChannels;
  wavHeader.wBitsPerSample=wBitsPerSample;
  hFilename=CreateWav(argv[argc-1],&wavHeader);

  /*
   * Record
   */
  SetDspDesc(hAudio,&wavHeader);
  printf("Recording (CNTL-C to stop)...\n");
  signal(SIGINT,stoprecord);
  while(dStop==0) {
    memset(aBuffer,0,AUDIO_BUFFER);
    j=read(hAudio,aBuffer,AUDIO_BUFFER);
    write(hFilename,aBuffer,j);
    dCount++;
  }
  dBytes=dCount*AUDIO_BUFFER;
  dSamples=dBytes/wavHeader.wBlockAlign;
  FixWav(hFilename,dSamples,dBytes);
  close(hFilename);

  exit(0);
}




void stoprecord(int signo)
{
  dStop++;
}
